File: codepen-frontend/src/types/group.ts
import type { User } from '@/types/user'
import type { Problem } from '@/types/problem'

// [신규] 그룹원 목록은 이제 User 정보에 "이 그룹에 언제 가입했는지"(joined_at)가 같이 옵니다.
export interface GroupMemberUser extends User {
  joined_at: string
}

export interface Group {
  group_id: number
  group_name: string
  description: string | null
  owner_id: string
  created_at: string
  owner: User
  members: GroupMemberUser[]
  problems: Problem[]
  invite_code: string
}

File: codepen-backend/models/group.py
from datetime import datetime, timezone
from sqlmodel import Field, SQLModel, Relationship
from sqlalchemy import Column, DateTime
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from .user import User
    from .group_member import GroupMember
    from .problem import Problem
    from .invite_queue import InviteQueue
    from .category import Category


class Group(SQLModel, table=True):
    group_id: int | None = Field(
        primary_key=True, unique=True, index=True, default=None
    )
    group_name: str = Field(nullable=False)
    description: str | None = Field(default=None, nullable=True)
    owner_id: str = Field(foreign_key="user.user_id", nullable=False)
    invite_code: str = Field(nullable=False, unique=True)

    created_at: datetime | None = Field(
        sa_column=Column(
            DateTime(timezone=True),
            nullable=False,
            default=lambda: datetime.now(timezone.utc),
        )
    )

    members: list["GroupMember"] = Relationship(back_populates="group")
    owner: "User" = Relationship(back_populates="owned_groups")
    problems: list["Problem"] = Relationship(back_populates="group")
    invite_queues: list["InviteQueue"] = Relationship(back_populates="group")
    categories: list["Category"] = Relationship(back_populates="group")
